diff --git a/test/test_jsc/test_runtime.py b/test/test_jsc/test_runtime.py
index abd8b9f3f..14fa43bfa 100644
--- a/test/test_jsc/test_runtime.py
+++ b/test/test_jsc/test_runtime.py
@@ -18,10 +18,8 @@
     SigChallengeInput,
     SigChallengeOutput,
 )
-from yt_dlp.extractor.youtube.jsc._builtin.bun import BunJCP
-from yt_dlp.extractor.youtube.jsc._builtin.deno import DenoJCP
-from yt_dlp.extractor.youtube.jsc._builtin.node import NodeJCP
 
+from yt_dlp_plugins.extractor.ytjsc import AppleWebKitJCP
 
 pytestmark = pytest.mark.skipif(not yt_dlp_ejs, reason='yt-dlp-ejs not available')
 
@@ -63,7 +61,7 @@
 ]
 
 
-@pytest.fixture(params=[BunJCP, DenoJCP, NodeJCP])
+@pytest.fixture(params=[AppleWebKitJCP])
 def jcp(request, ie, logger):
     obj = request.param(ie, logger, None)
     if not obj.is_available():
